File: apps/apple-watch/IslamicDayDialWatch/IslamicDayCore/Ring.swift
import Foundation

/// Ring geometry: progress, markers, segments.
/// Mirrors packages/core/src/ring.ts

private let MARKER_DEFS: [(String, KeyPath<ComputedTimeline, Date>, RingMarkerKind)] = [
    ("maghrib", \.lastMaghrib, .primary),
    ("isha", \.isha, .primary),
    ("last_third_start", \.lastThirdStart, .secondary),
    ("fajr", \.fajr, .primary),
    ("sunrise", \.sunrise, .primary),
    ("dhuhr", \.dhuhr, .primary),
    ("asr", \.asr, .primary),
]

private let SEGMENT_BOUNDARIES: [(IslamicPhaseId, KeyPath<ComputedTimeline, Date>, KeyPath<ComputedTimeline, Date>)] = [
    (.maghrib_to_isha, \.lastMaghrib, \.isha),
    (.isha_to_midnight, \.isha, \.lastThirdStart),
    (.last_third_to_fajr, \.lastThirdStart, \.fajr),
    (.fajr_to_sunrise, \.fajr, \.sunrise),
    (.sunrise_to_dhuhr, \.sunrise, \.dhuhr),
    (.dhuhr_to_asr, \.dhuhr, \.asr),
    (.asr_to_maghrib, \.asr, \.nextMaghrib),
]

func getIslamicDayProgress(now: Date, lastMaghrib: Date, nextMaghrib: Date) -> Double {
    let total = nextMaghrib.timeIntervalSince(lastMaghrib)
    if total <= 0 { return 0 }
    let elapsed = now.timeIntervalSince(lastMaghrib)
    return min(1, max(0, elapsed / total))
}

func timestampToAngle(timestamp: Date, lastMaghrib: Date, nextMaghrib: Date) -> Double {
    return getIslamicDayProgress(now: timestamp, lastMaghrib: lastMaghrib, nextMaghrib: nextMaghrib) * 360
}

func getMarkers(timeline: ComputedTimeline) -> [RingMarker] {
    MARKER_DEFS.map { (id, key, kind) in
        let ts = timeline[keyPath: key]
        return RingMarker(
            id: id,
            timestamp: ts,
            angleDeg: timestampToAngle(timestamp: ts, lastMaghrib: timeline.lastMaghrib, nextMaghrib: timeline.nextMaghrib),
            kind: kind
        )
    }
}

func getRingSegments(timeline: ComputedTimeline) -> [RingSegment] {
    SEGMENT_BOUNDARIES.map { (id, startKey, endKey) in
        let start = timeline[keyPath: startKey]
        let end = timeline[keyPath: endKey]
        return RingSegment(
            id: id,
            start: start,
            end: end,
            startAngleDeg: timestampToAngle(timestamp: start, lastMaghrib: timeline.lastMaghrib, nextMaghrib: timeline.nextMaghrib),
            endAngleDeg: timestampToAngle(timestamp: end, lastMaghrib: timeline.lastMaghrib, nextMaghrib: timeline.nextMaghrib)
        )
    }
}

File: packages/core/src/ring.ts
import type {
  ComputedTimeline,
  RingMarker,
  RingSegment,
  RingMarkerId,
  RingMarkerKind,
} from './types.js';
import { PHASE_BOUNDARIES } from './types.js';

/**
 * Linear progress through the Islamic day: 0 at lastMaghrib, 1 at nextMaghrib.
 */
export function getIslamicDayProgress(
  now: Date,
  lastMaghrib: Date,
  nextMaghrib: Date,
): number {
  const total = nextMaghrib.getTime() - lastMaghrib.getTime();
  if (total <= 0) return 0;
  const elapsed = now.getTime() - lastMaghrib.getTime();
  return Math.max(0, Math.min(1, elapsed / total));
}

/**
 * Convert a timestamp to an angle (0–360°) on the ring.
 * 0° = top = lastMaghrib, angles increase clockwise.
 */
export function timestampToAngle(
  timestamp: Date,
  lastMaghrib: Date,
  nextMaghrib: Date,
): number {
  const progress = getIslamicDayProgress(timestamp, lastMaghrib, nextMaghrib);
  return progress * 360;
}

const MARKER_DEFS: ReadonlyArray<{
  id: RingMarkerId;
  timelineKey: keyof ComputedTimeline;
  kind: RingMarkerKind;
}> = [
  { id: 'maghrib', timelineKey: 'lastMaghrib', kind: 'primary' },
  { id: 'isha', timelineKey: 'isha', kind: 'primary' },
  { id: 'last_third_start', timelineKey: 'lastThirdStart', kind: 'secondary' },
  { id: 'fajr', timelineKey: 'fajr', kind: 'primary' },
  { id: 'sunrise', timelineKey: 'sunrise', kind: 'primary' },
  { id: 'duha_start', timelineKey: 'duhaStart', kind: 'secondary' },
  { id: 'duha_end', timelineKey: 'duhaEnd', kind: 'secondary' },
  { id: 'dhuhr', timelineKey: 'dhuhr', kind: 'primary' },
  { id: 'asr', timelineKey: 'asr', kind: 'primary' },
];

export function getMarkers(timeline: ComputedTimeline): RingMarker[] {
  return MARKER_DEFS.map(({ id, timelineKey, kind }) => {
    const timestamp = timeline[timelineKey];
    return {
      id,
      timestamp,
      angleDeg: timestampToAngle(timestamp, timeline.lastMaghrib, timeline.nextMaghrib),
      kind,
    };
  });
}

export function getRingSegments(timeline: ComputedTimeline): RingSegment[] {
  return PHASE_BOUNDARIES.map(({ id, startKey, endKey }) => {
    const start = timeline[startKey];
    const end = timeline[endKey];
    return {
      id,
      start,
      end,
      startAngleDeg: timestampToAngle(start, timeline.lastMaghrib, timeline.nextMaghrib),
      endAngleDeg: timestampToAngle(end, timeline.lastMaghrib, timeline.nextMaghrib),
    };
  });
}
